File: Sources/RichTextView/Serialization/RichTextMarkdownSerializer.swift
// Sources/RichTextView/Serialization/RichTextMarkdownSerializer.swift
//
// Model -> Markdown. The inverse of the parser; used for plain round-trip on the pasteboard.

import Foundation

public enum RichTextMarkdownSerializer {

    public static func string(from document: RichTextDocument) -> String {
        return blocks(document.blocks, indent: "")
    }

    private static func blocks(_ blocks: [RichTextBlock], indent: String) -> String {
        return blocks.map { block(_:$0, indent: indent) }.joined(separator: "\n\n")
    }

    private static func block(_ block: RichTextBlock, indent: String) -> String {
        switch block {
        case .heading(let level, let inlines):
            return String(repeating: "#", count: level) + " " + inline(inlines)
        case .paragraph(let inlines):
            return inline(inlines)
        case .codeBlock(let language, let code):
            return "```" + (language ?? "") + "\n" + code + "\n```"
        case .blockQuote(let inner):
            return blocks(inner, indent: "").split(separator: "\n", omittingEmptySubsequences: false)
                .map { "> " + $0 }.joined(separator: "\n")
        case .list(let ordered, let start, _, let items):
            return list(ordered: ordered, start: start, items: items, indent: indent)
        case .thematicBreak:
            return "---"
        case .table(let headers, let alignments, let rows):
            return table(headers: headers, alignments: alignments, rows: rows)
        }
    }

    private static func list(ordered: Bool, start: Int, items: [[RichTextBlock]], indent: String) -> String {
        var lines: [String] = []
        for (offset, itemBlocks) in items.enumerated() {
            let marker = ordered ? "\(start + offset). " : "- "
            let rendered = blocks(itemBlocks, indent: indent + "  ")
            let parts = rendered.split(separator: "\n", omittingEmptySubsequences: false)
            for (lineIndex, part) in parts.enumerated() {
                if lineIndex == 0 {
                    lines.append(indent + marker + part)
                } else {
                    lines.append(indent + "  " + part)
                }
            }
        }
        return lines.joined(separator: "\n")
    }

    private static func table(headers: [[RichTextInline]], alignments: [RichTextColumnAlignment],
                              rows: [[[RichTextInline]]]) -> String {
        let columns = max(headers.count, rows.map(\.count).max() ?? 0)
        func cell(_ cells: [[RichTextInline]], _ c: Int) -> String {
            return c < cells.count ? inline(cells[c]) : ""
        }
        func delimiter(_ c: Int) -> String {
            let a = c < alignments.count ? alignments[c] : .none
            switch a {
            case .left: return ":---"
            case .center: return ":---:"
            case .right: return "---:"
            default: return "---"
            }
        }
        var lines: [String] = []
        lines.append("| " + (0..<columns).map { cell(headers, $0) }.joined(separator: " | ") + " |")
        lines.append("| " + (0..<columns).map(delimiter).joined(separator: " | ") + " |")
        for row in rows {
            lines.append("| " + (0..<columns).map { cell(row, $0) }.joined(separator: " | ") + " |")
        }
        return lines.joined(separator: "\n")
    }

    static func inline(_ nodes: [RichTextInline]) -> String {
        var out = ""
        for node in nodes {
            switch node {
            case .text(let s):
                out += s
            case .emphasis(let c):
                out += "*" + inline(c) + "*"
            case .strong(let c):
                out += "**" + inline(c) + "**"
            case .strikethrough(let c):
                out += "~~" + inline(c) + "~~"
            case .code(let s):
                out += "`" + s + "`"
            case .link(let c, let url):
                out += "[" + inline(c) + "](" + url + ")"
            case .image(let alt, let url):
                out += "![" + alt + "](" + url + ")"
            case .lineBreak:
                out += "  \n"
            }
        }
        return out
    }
}

File: android/richtext/src/main/kotlin/com/abracode/richtext/serialization/RichTextMarkdownSerializer.kt
package com.abracode.richtext.serialization

import com.abracode.richtext.model.RichTextBlock
import com.abracode.richtext.model.RichTextColumnAlignment
import com.abracode.richtext.model.RichTextDocument
import com.abracode.richtext.model.RichTextInline

// Model -> Markdown. The inverse of the parser; used for plain round-trip on the clipboard. A 1:1 port of
// Sources/RichText/Serialization/RichTextMarkdownSerializer.swift.

object RichTextMarkdownSerializer {

    fun string(document: RichTextDocument): String = blocks(document.blocks, "")

    private fun blocks(blocks: List<RichTextBlock>, indent: String): String =
        blocks.joinToString("\n\n") { block(it, indent) }

    private fun block(block: RichTextBlock, indent: String): String = when (block) {
        is RichTextBlock.Heading ->
            "#".repeat(block.level) + " " + inline(block.inlines)
        is RichTextBlock.Paragraph ->
            inline(block.inlines)
        is RichTextBlock.CodeBlock ->
            "```" + (block.language ?: "") + "\n" + block.code + "\n```"
        is RichTextBlock.BlockQuote ->
            blocks(block.blocks, "").split("\n").joinToString("\n") { "> $it" }
        is RichTextBlock.ListBlock ->
            list(block.ordered, block.start, block.items, indent)
        RichTextBlock.ThematicBreak ->
            "---"
        is RichTextBlock.Table ->
            table(block.headers, block.alignments, block.rows)
    }

    private fun list(ordered: Boolean, start: Int, items: List<List<RichTextBlock>>, indent: String): String {
        val lines = mutableListOf<String>()
        for ((offset, itemBlocks) in items.withIndex()) {
            val marker = if (ordered) "${start + offset}. " else "- "
            val rendered = blocks(itemBlocks, indent + "  ")
            val parts = rendered.split("\n")
            for ((lineIndex, part) in parts.withIndex()) {
                if (lineIndex == 0) {
                    lines.add(indent + marker + part)
                } else {
                    lines.add(indent + "  " + part)
                }
            }
        }
        return lines.joinToString("\n")
    }

    private fun table(
        headers: List<List<RichTextInline>>,
        alignments: List<RichTextColumnAlignment>,
        rows: List<List<List<RichTextInline>>>,
    ): String {
        val columns = maxOf(headers.size, rows.maxOfOrNull { it.size } ?: 0)
        fun cell(cells: List<List<RichTextInline>>, c: Int): String = if (c < cells.size) inline(cells[c]) else ""
        fun delimiter(c: Int): String {
            val a = if (c < alignments.size) alignments[c] else RichTextColumnAlignment.NONE
            return when (a) {
                RichTextColumnAlignment.LEFT -> ":---"
                RichTextColumnAlignment.CENTER -> ":---:"
                RichTextColumnAlignment.RIGHT -> "---:"
                RichTextColumnAlignment.NONE -> "---"
            }
        }
        val lines = mutableListOf<String>()
        lines.add("| " + (0 until columns).joinToString(" | ") { cell(headers, it) } + " |")
        lines.add("| " + (0 until columns).joinToString(" | ") { delimiter(it) } + " |")
        for (row in rows) {
            lines.add("| " + (0 until columns).joinToString(" | ") { cell(row, it) } + " |")
        }
        return lines.joinToString("\n")
    }

    fun inline(nodes: List<RichTextInline>): String {
        val out = StringBuilder()
        for (node in nodes) {
            when (node) {
                is RichTextInline.Text -> out.append(node.value)
                is RichTextInline.Emphasis -> out.append("*").append(inline(node.children)).append("*")
                is RichTextInline.Strong -> out.append("**").append(inline(node.children)).append("**")
                is RichTextInline.Strikethrough -> out.append("~~").append(inline(node.children)).append("~~")
                is RichTextInline.Code -> out.append("`").append(node.value).append("`")
                is RichTextInline.Link -> out.append("[").append(inline(node.text)).append("](").append(node.url).append(")")
                is RichTextInline.Image -> out.append("![").append(node.alt).append("](").append(node.url).append(")")
                RichTextInline.LineBreak -> out.append("  \n")
            }
        }
        return out.toString()
    }
}
